PEP 724: Stricter TypeGuard (#3266)
python/peps: pep-0724.rst

PEP: 724
Title: Stricter Type Guards
Author: Rich Chiodo <rchiodo at microsoft.com>, Eric Traut <erictr at microsoft.com>, Erik De Bonte <erikd at microsoft.com>
Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
Status: Draft
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 28-Jul-2023
Python-Version: 3.13
Post-History: 

> Resolution: 

Reviewer: To satisfy the linter
```suggestion
```

Abstract
========

:pep:`647` introduced the concept of ``TypeGuard`` functions which return true
if their input parameter matches their target type. For example, a function that

> returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's

Reviewer 2: ``true`` should be ``True`` to match the name in Python.

> returns ``TypeGuard[str]`` is assumed to return ``true`` if and only if it's

Reviewer 2: `it's` should be `its`

> input parameter is a ``str``. This allows type checkers to narrow types in this

Reviewer 2: `this positive case` should be `the positive case`. You should also explain what "positive case" and "negative case" mean before using those terms.

positive case.

This PEP further refines :pep:`647` by allowing type checkers to also narrow types

> when a ``TypeGuard`` function returns false.

Author: Oh I thought I was. I'm using a tool that autowraps to some column with. Maybe it's 80.

Motivation
==========

``TypeGuards`` are used throughout Python libraries to allow a type checker to
narrow the type of something when the ``TypeGuard`` returns true.

However, in the ``else`` clause, :pep:`647` didn't prescribe what the type might
be:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type here is not narrowed. It is still 'str | int' 

> This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype

Reviewer 3: This should be in the specification, not the motivation.

> This PEP proposes that when the type argument of the ``TypeGuard`` is a subtype

Reviewer 2: The term "subtype" is probably not sufficient here. Subtype relationships are well-defined in the Python type system as long as there is no `Any` involved. As soon as you introduce an `Any`, you can no longer talk about subtyping. Instead, you need to talk about "type consistency" or "consistent types". See PEP 483 for a discussion of how this differs.

For this spec, I should use "consistent types" rather than "subtypes" as the test. That's how I've coded the current prototype.

Author: Didn't realize there was a 'is-consistent' term that deals with Any. I'll update this text and add a footnote about PEP 483

of the type of the first input parameter, then the false return case can be
further narrowed. 

This changes the example above like so:

.. code-block:: python

    def is_str(val: str | int) -> TypeGuard[str]:
        return isinstance(val, str)

    def func(val: str | int):
        if is_str(val):
            # Type checkers can assume val is a 'str' in this branch
        else:
            # Type checkers can assume val is an 'int' in this branch

Since the ``TypeGuard`` type (or output type) is a subtype of the first input
parameter type, a type checker can determine that the only possible type in the
``else`` is the other type in the ``Union``. In this example, it is safe to
assume that if ``is_str`` returns false, then type of the ``val`` argument is an
``int``.

> Specification

Reviewer 2: The spec talks about type narrowing in the negative case (which is part of what "strict type guard semantics" provide), but it doesn't talk about narrowing of unions based on constraints imposed by the type guard return type (which is the second part of what strict type guard semantics provide). For details, refer to my original post [here](https://github.com/python/typing/discussions/1013#discussioncomment-1966238). In particular, pay close attention to point 3.

I recommend that you split the specification into two part:
1. The conditions under which "strict type guard semantics" will be applied. This is effectively point 1 from my original post (but instead of enforcing the requirement, it uses this as a condition as a test for whether strict semantics can be applied).
2. A detailed description of what is implied by "strict type guard semantics". This should include both points 2 and 3 from my original post.

definition of ``TypeGuard`` for type checkers. Runtimes are already behaving
in this way.

Existing ``TypeGuard`` usage may change though, as described in 
`Backwards Compatibility`_.

Unsafe Narrowing
--------------------

There are cases where this further type narrowing is not possible though. Here's
an example:

.. code-block:: python

    def is_str_list(val: list[int | str]) -> TypeGuard[list[str]]
        return all(isinstance(x, str) for x in val)

    def func(val: list[int | str]):
        if is_str_list(val):
            # Type checker assumes list[str] here
        else:
            # Type checker cannot assume list[int] here

Since ``list`` is invariant, it doesn't have any subtypes. ``list[str]`` is not
a subtype of ``list[str | int]``. This means type checkers cannot narrow the
type to ``list[int]`` in the false case.

Type checkers should not assume any narrowing in the false case when the
``TypeGuard`` type argument is not a subtype of the first input parameter type. 

However, narrowing in the true case is still possible. In the example above, the
type checker can assume the list is a ``list[str]`` if the ``TypeGuard``

> function returns true.

Author: Yes this is what PEP 647 stated. I'll add something to say that.

User error
--------------------------

The new ``else`` case for a ``TypeGuard`` can be setup incorrectly. Here's an
example:

.. code-block:: python

    def is_positive_int(val: int | str) -> TypeGuard[int]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):
            # Type checker assumes int here
        else:
            # Type checker assumes str here

A type checker will assume for the else case that the value is ``str``. This is
a change in behavior from :pep:`647` but as that pep stated, there are many ways
a determined or uninformed developer can subvert type safety.

A better way to handle this example would be something like so:

.. code-block:: python

    PosInt = NewType('PosInt', int)

    def is_positive_int(val: PosInt | int | str) -> TypeGuard[PosInt]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):
            # Type checker assumes PosInt here
        else:
            # Type checker assumes str | int here

Backwards Compatibility
=======================

> For preexisting code this PEP should require no changes.

Reviewer: ```suggestion
For preexisting code, this PEP should require no changes.
```

However, some use cases such as the one below can be simplified:

.. code-block:: python

    class A():
        pass
    class B():
        pass

    def is_A(x: A | B) -> TypeGuard[A]:
        return is_instance(x, A)

    def is_B(x: A | B) -> TypeGuard[B]:
        return is_instance(x, B)

    def test(x: A | B):
        if is_A(x):
            # Do stuff assuming x is an 'A'
            return
        assert is_B(x)

        # Do stuff assuming x is a 'B'
        return

> 
> With this proposed change, the code above continues to work but could be
> simplified by removing the assertion that x is of type B in the negative case:

Reviewer: ```suggestion
With this proposed change, the code above continues to work but could be
simplified by removing the assertion that ``x`` is of type ``B`` in the negative case:
```

.. code-block:: python

    class A():
        pass
    class B():
        pass

    def is_A(x: A | B) -> TypeGuard[A]:
        return is_instance(x, A)

    def test(x: A | B):
        if is_A(x):
            # Do stuff assuming x is an 'A'
            return

        # Do stuff assuming x is a 'B'
        return

How to Teach This
=================

> The belief is that new users will assume this is how ``TypeGuard`` works in the
> first place. Meaning this change should make ``TypeGuard`` easier to teach.

Reviewer: A suggested re-wording which makes it clearer that this is an assertion / belief on the part of the authors, and that I think clarifies the intent here. 

```suggestion
We assert that users unfamiliar with ``TypeGuard`` will expect
the behavior outlined in this PEP,
therefore making ``TypeGuard`` easier to teach and explain.
```

Reference Implementation
========================

> A reference `implementation <https://github.com/microsoft/pyright/commit/9a5af798d726bd0612cebee7223676c39cf0b9b0>`__ of this idea exists in Pyright.

Reviewer: ```suggestion
A reference `implementation`__ of this idea exists in Pyright.

__ https://github.com/microsoft/pyright/commit/9a5af798d726bd0612cebee7223676c39cf0b9b0
```

Rejected Ideas
==============

Originally a new ``StrictTypeGuard`` construct was proposed. A
``StrictTypeGuard`` would be similar to to a ``TypeGuard`` except it would
explicitly state that output type was a subtype of the input type. Type checkers
would validate that the output type was a subtype of the input type.

> See this comment: `StrictTypeGuard proposal <https://github.com/python/typing/discussions/1013#discussioncomment-1966238>`__

Reviewer: ```suggestion
See this comment: `StrictTypeGuard proposal`__

__ https://github.com/python/typing/discussions/1013#discussioncomment-1966238
```

PEP: 724
Title: Stricter Type Guards
Author: Rich Chiodo <rchiodo at microsoft.com>, 
        Eric Traut <erictr at microsoft.com>, 
        Erik De Bonte <erikd at microsoft.com>
Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
Status: Draft
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 28-Jul-2023
Python-Version: 3.13
Post-History: 

Abstract
========

:pep:`647` introduced the concept of ``TypeGuard`` functions which return
``True`` if their input parameter matches their target type. For example, a
function that returns ``TypeGuard[str]`` is assumed to return ``True`` if and
only if its input parameter is a ``str``. This allows type checkers to narrow
types when a ``TypeGuard`` function returns ``True``.

This PEP further refines :pep:`647` by allowing type checkers to also narrow
types when a ``TypeGuard`` function returns ``False``.

Motivation
==========

> ``TypeGuards`` are used throughout Python libraries to allow a type checker to

Reviewer 4: How to render as "`TypeGuard`s" in RST:

```suggestion
``TypeGuard``\s are used throughout Python libraries to allow a type checker to
```

Covariance
----------

Covariance is not the same thing as being consistent [#isconsistent]_:

.. code-block:: python

    def is_str_seq(val: Sequence[str | int]) -> TypeGuard[Sequence[str]]:
        return all(isinstance(x, str) for x in val)

    def func3(val: Sequence[str | int]):
        if is_str_seq(val):
            # val is Sequence[str]
        else:
            # val is Sequence[str | int]

Type ``B`` is consistent [#isconsistent]_ with type ``A`` if

* ``B`` is a subtype of ``A``
* ``A`` is Any
* ``B`` is Any

``Sequence[str]`` is not a subtype of ``Sequence[str | int]`` so the special
semantics cannot be applied and only the ``True`` case is narrowed. 

Degenerate case
---------------

It is possible to break the type checker by making a ``TypeGuard`` that is
incorrect:

.. code-block:: python

    def is_positive_int(val: int | str) -> TypeGuard[int]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):
            # Type checker assumes int here
        else:
            # Type checker assumes str incorrectly here

As :pep:`647 <647#enforcing-strict-narrowing>` stated, users can create invalid

> ``TypeGuards``. The same thing applies to this PEP. If the user is not careful

Reviewer 4: ```suggestion
``TypeGuard``\s. The same thing applies to this PEP. If the user is not careful
```

PEP: 724
Title: Stricter Type Guards
Author: Rich Chiodo <rchiodo at microsoft.com>, 
        Eric Traut <erictr at microsoft.com>, 
        Erik De Bonte <erikd at microsoft.com>
Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Discussions-To: https://mail.python.org/archives/list/typing-sig@python.org/thread/EMUD2D424OI53DCWQ4H5L6SJD2IXBHUL/
Status: Draft
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 28-Jul-2023
Python-Version: 3.13
Post-History: 

> Resolution: 

Reviewer: ```suggestion
```

Reviewer: @rchiodo -- FYI, the lint checks will fail with the 'resolution' header blank -- we expect it to be missing initially, then filled in with a link to a resolution post (once your PEP is accepted, hopefully!).

A

        else:
            # Type checkers can assume val is an 'int' in this branch

* When the output type of the ``TypeGuard`` is a union, the type checker can
  apply additional type narrowing based on the type of the first input
  argument, eliminating union elements that are impossible given the input
  argument type:

.. code-block:: python

    def is_cardinal_direction(val: str) -> TypeGuard[Literal["N", "S", "E", "W"]]:
        return val in ("N", "S", "E", "W")

    def func(direction: Literal["NW", "E"]):
        if is_cardinal_direction(direction):
            # Literal["E"] The type cannot be "N", "S" or "W" here because of
            # argument type
        else:
            # Literal["NW"]

This logic can be summed up in a table:

============ ======================= ===================
|space|      Non strict TypeGuard    Strict TypeGuard
============ ======================= ===================
Applies when R not consistent with P R consistent with P
NP is ..     R                       A & R
NN is ..     A                       A & ~R
============ ======================= ===================

The definition of ``A & R`` and ``A & ~R`` would be left up to 

> the type checker itself as these concepts don't yet exist in Python typing.

Reviewer 5: We're currently working on a definition of intersections prior to a PEP, and the question of Not also has come up over there. I would strongly suggest not leaving Not fully up to type checkers here\*, and to write this in a way where This should defer to a future intersection pep if this is ready before the intersection pep is. The intersection pep's current track is consistent with the behavior laid out as desirable here.

\* [This comment goes into some detail about what Not can be shown to be safely composed with.](https://github.com/python/typing/issues/801#issuecomment-1679883519 ) Notably, it hasn't been shown that type negation is safe to apply to gradual or unknown types. This PEP does not currently suggest to do this or allow this, but it should probably not introduce an internal type negation without forbiding type-checkers from expanding on this to prevent accidentally introducing something with unclear meaning in the process.

Author: Thanks that sounds like a much better idea than just leaving this vague. I'll reference the intersection design doc.

Reviewer 2: I don't think a draft PEP should reference an incomplete design document. If and when the Intersection design reaches the point where it is submitted as a PEP, then it would be fine to provide a footnote in this PEP that references it. In the meantime, I think it's fine to say something to the effect that "future extensions to the type system may affect the implementation of this feature".

Author: I changed the wording but still left at least a potential for what might the change might be?

Reviewer 5: I agree with Eric that it probably shouldn't reference the incomplete design, I only provided the link to be able to have it be confirmed that the work on intersections matches how you're using it here and for anyone who might be interested in exploring how the two in progress PEPs might interact in case people had concerns or additional feedback to either draft as a result.

Intersection of types and type negation are not defined and are left up to the
type checker to decide on how to implement. Future `extensions`__ to the type
system may change this behavior though.

__ https://github.com/CarliJoy/intersection_examples/blob/main/specification.rst

.. |space| unicode:: 0x2008

Additional Examples
===================

Any
---

``Any`` is consistent [#isconsistent]_ with any other type, which means
stricter semantics can be applied. 

.. code-block:: python

    def is_a(x: Any) -> TypeGuard[A]: # Stricter TypeGuard possible
        return isinstance(x, A)

    def test(x: A | B):
        if is_a(x):
            # x is of type A here
        else:
            # x is of type B here

Degenerate case
---------------

It is possible to break the type checker by making a ``TypeGuard`` that is
incorrect:

.. code-block:: python

    def is_positive_int(val: int | str) -> TypeGuard[int]:
        return isinstance(val, int) and val > 0

    def func(val: int | str):
        if is_positive_int(val):
            # Type checker assumes int here
        else:
            # Type checker assumes str incorrectly here

> As :pep:`647 <647#enforcing-strict-narrowing>` stated, users can create invalid

Reviewer 3: Worth noting that under PEP 647, this particular TypeGuard is actually correct and meaningful. I still think that's worth calling out explicitly in the "Backwards Compatibility" section, though that section is now much better than when I last read the PEP.

Specification
=============

Given a ``TypeGuard`` and a calling function like so:

.. code-block:: python

    def guard(x: P) -> TypeGuard[R]: ...

    def func1(val: A):
        if guard(val):
            reveal_type(val) # NP
        else:
            reveal_type(val) # NN

Each use of a ``TypeGuard`` involves five types:

* P = ``TypeGuard`` parameter type

* R = ``TypeGuard`` return type

* A = Argument type (pre-narrowed)

* NP = Narrowed type (positive)

> * NN = Narrowed type (negative)

Reviewer: You don't need a blank line in-between each item of a list just to confirm -- but this might be intentional!

        else:
            # Literal["NW"]

This logic can be summed up in a table:

============ ======================= ===================
|space|      Non strict TypeGuard    Strict TypeGuard
============ ======================= ===================
Applies when R not consistent with P R consistent with P
NP is ..     R                       A & R

> NN is ..     A                       A & ~R

Reviewer: Should we put these in literal markup (or maths markup?)

Reviewer: * https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#math
* https://www.docutils.org/docs/ref/rst/roles.html#math
* https://www.docutils.org/docs/ref/rst/directives.html#math

``:math`...` `` should work, ``.. math:: ...`` for display maths.

Author: Thanks for the help. Not sure if that looks better but it might be more consistent with other PEPs that talk about set logic.
